TestResourceException: Invalid status line

### Expected Behavior

Successfully run test classes annotated with `@MicronautTest` when Micronaut Test Resource Server is running.

Note:
This used to work for me with no issue. Since then, I haven't made dependency updates. Only code updates. I also have issues starting/stopping the test resource service. Any help in knowing how to find/kill the test resource process is appreciated.

### Actual Behaviour

When I run my tests annotated with `@MicronautTest` via Maven commands or IntelliJ, I receive the following error:
```
io.micronaut.testresources.client.TestResourcesException: java.io.IOException: Invalid status line: "      A{"Dns4":{"ProtectionState":"open"},"Ip4":{"ProtectionState":"notProtected","connectionError":"false"},"Dns6":{"ProtectionState":"open"},"Ip6":{"ProtectionState":"notProtected","connectionError":"false"},"LastConnectedTime":"0","LogUploadEnabled":"false","ClientName":"UnknownClient","UserName":"UnknownUser","AgentStatus":"noOrginfo","SWG":{"State":"unprotected","Licensed":"false","Enabled":"false"}}"

	at io.micronaut.testresources.client.DefaultTestResourcesClient.request(DefaultTestResourcesClient.java:157)
	at io.micronaut.testresources.client.DefaultTestResourcesClient.getRequiredPropertyEntries(DefaultTestResourcesClient.java:104)
	at java.base/java.util.Optional.map(Optional.java:260)
	at io.micronaut.testresources.client.TestResourcesClientPropertySourceLoader$ClientTestResourcesResolver.getPropertyEntries(TestResourcesClientPropertySourceLoader.java:50)
	at io.micronaut.testresources.core.LazyTestResourcesPropertySourceLoader$LazyPropertySource.computeKeys(LazyTestResourcesPropertySourceLoader.java:98)
	at io.micronaut.testresources.core.LazyTestResourcesPropertySourceLoader$LazyPropertySource.iterator(LazyTestResourcesPropertySourceLoader.java:90)
	at io.micronaut.context.env.PropertySourcePropertyResolver.processPropertySource(PropertySourcePropertyResolver.java:588)
	at io.micronaut.context.env.DefaultEnvironment.readPropertySources(DefaultEnvironment.java:459)
	at io.micronaut.context.env.DefaultEnvironment.start(DefaultEnvironment.java:277)
	at io.micronaut.context.DefaultApplicationContext$RuntimeConfiguredEnvironment.start(DefaultApplicationContext.java:909)
	at io.micronaut.context.DefaultApplicationContext$RuntimeConfiguredEnvironment.start(DefaultApplicationContext.java:878)
	at io.micronaut.context.DefaultApplicationContext.startEnvironment(DefaultApplicationContext.java:266)
	at io.micronaut.context.DefaultApplicationContext.start(DefaultApplicationContext.java:201)
	at io.micronaut.test.extensions.AbstractMicronautExtension.startApplicationContext(AbstractMicronautExtension.java:507)
	at io.micronaut.test.extensions.AbstractMicronautExtension.beforeClass(AbstractMicronautExtension.java:346)
	at io.micronaut.test.extensions.junit5.MicronautJunit5Extension.beforeAll(MicronautJunit5Extension.java:84)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
Caused by: java.io.IOException: Invalid status line: "      A{"Dns4":{"ProtectionState":"open"},"Ip4":{"ProtectionState":"notProtected","connectionError":"false"},"Dns6":{"ProtectionState":"open"},"Ip6":{"ProtectionState":"notProtected","connectionError":"false"},"LastConnectedTime":"0","LogUploadEnabled":"false","ClientName":"UnknownClient","UserName":"UnknownUser","AgentStatus":"noOrginfo","SWG":{"State":"unprotected","Licensed":"false","Enabled":"false"}}"
	at java.net.http/jdk.internal.net.http.HttpClientImpl.send(HttpClientImpl.java:586)
	at java.net.http/jdk.internal.net.http.HttpClientFacade.send(HttpClientFacade.java:119)
	at io.micronaut.testresources.client.DefaultTestResourcesClient.request(DefaultTestResourcesClient.java:140)
	... 16 more
Caused by: java.net.ProtocolException: Invalid status line: "      A{"Dns4":{"ProtectionState":"open"},"Ip4":{"ProtectionState":"notProtected","connectionError":"false"},"Dns6":{"ProtectionState":"open"},"Ip6":{"ProtectionState":"notProtected","connectionError":"false"},"LastConnectedTime":"0","LogUploadEnabled":"false","ClientName":"UnknownClient","UserName":"UnknownUser","AgentStatus":"noOrginfo","SWG":{"State":"unprotected","Licensed":"false","Enabled":"false"}}"
	at java.net.http/jdk.internal.net.http.Http1HeaderParser.protocolException(Http1HeaderParser.java:342)
	at java.net.http/jdk.internal.net.http.Http1HeaderParser.readStatusLineFeed(Http1HeaderParser.java:190)
	at java.net.http/jdk.internal.net.http.Http1HeaderParser.parse(Http1HeaderParser.java:121)
	at java.net.http/jdk.internal.net.http.Http1Response$HeadersReader.handle(Http1Response.java:690)
	at java.net.http/jdk.internal.net.http.Http1Response$HeadersReader.handle(Http1Response.java:621)
	at java.net.http/jdk.internal.net.http.Http1Response$Receiver.accept(Http1Response.java:612)
	at java.net.http/jdk.internal.net.http.Http1Response$HeadersReader.tryAsyncReceive(Http1Response.java:668)
	at java.net.http/jdk.internal.net.http.Http1AsyncReceiver.flush(Http1AsyncReceiver.java:233)
	at java.net.http/jdk.internal.net.http.common.SequentialScheduler$LockingRestartableTask.run(SequentialScheduler.java:205)
	at java.net.http/jdk.internal.net.http.common.SequentialScheduler$CompleteRestartableTask.run(SequentialScheduler.java:149)
	at java.net.http/jdk.internal.net.http.common.SequentialScheduler$SchedulableTask.run(SequentialScheduler.java:230)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

### Steps To Reproduce

1. Create test class and annotate with `@MicronautTest`
2. Create test
3. Start test service via `mvn mn:start-testresources-service`
4. Run test
5. See error

### Environment Information

- OS: Mac
- Maven
- Java 17
- Various Micronaut Dependency Versions
```
[INFO] +- io.micronaut.data:micronaut-data-hibernate-jpa:jar:4.7.0:compile
[INFO] |  +- io.micronaut:micronaut-inject:jar:4.4.9:compile
[INFO] |  |  \- io.micronaut:micronaut-core:jar:4.4.9:compile
[INFO] |  +- io.micronaut.data:micronaut-data-runtime:jar:4.7.1:compile
[INFO] |  +- io.micronaut.data:micronaut-data-model:jar:4.7.1:compile
[INFO] |  +- io.micronaut.data:micronaut-data-jpa:jar:4.7.1:compile
[INFO] +- io.micronaut.sql:micronaut-hibernate-jpa:jar:5.6.0:compile
[INFO] |  +- io.micronaut.validation:micronaut-validation:jar:4.5.0:compile
[INFO] |  |  \- io.micronaut.reactor:micronaut-reactor:jar:3.3.0:runtime
[INFO] |  +- io.micronaut:micronaut-aop:jar:4.4.9:compile
[INFO] |  \- io.micronaut.sql:micronaut-jdbc:jar:5.6.0:compile
[INFO] +- io.micronaut.data:micronaut-data-tx-hibernate:jar:4.7.1:compile
[INFO] |  +- io.micronaut.data:micronaut-data-tx:jar:4.7.1:compile
[INFO] |  |  \- io.micronaut.data:micronaut-data-connection:jar:4.7.1:compile
[INFO] |  +- io.micronaut.data:micronaut-data-tx-jdbc:jar:4.7.1:compile
[INFO] |  |  \- io.micronaut.data:micronaut-data-connection-jdbc:jar:4.7.1:compile
[INFO] |  \- io.micronaut.data:micronaut-data-connection-hibernate:jar:4.7.1:compile
[INFO] +- io.micronaut.sql:micronaut-jdbc-hikari:jar:5.6.0:compile
[INFO] +- io.micronaut.beanvalidation:micronaut-hibernate-validator:jar:4.3.0:compile
[INFO] +- io.micronaut.flyway:micronaut-flyway:jar:7.2.0:compile
[INFO] |  +- io.micronaut:micronaut-management:jar:4.4.9:compile
[INFO] |  |  \- io.micronaut:micronaut-router:jar:4.4.9:compile
[INFO] +- io.micronaut.serde:micronaut-serde-jackson:jar:2.9.0:compile
[INFO] |  +- io.micronaut:micronaut-jackson-core:jar:4.4.9:compile
[INFO] |  +- io.micronaut:micronaut-context:jar:4.4.9:compile
[INFO] |  +- io.micronaut.serde:micronaut-serde-api:jar:2.9.0:compile
[INFO] |  \- io.micronaut.serde:micronaut-serde-support:jar:2.9.0:runtime
[INFO] +- io.micronaut.servlet:micronaut-http-server-tomcat:jar:4.7.0:compile
[INFO] |  +- io.micronaut.servlet:micronaut-servlet-engine:jar:4.7.0:compile
[INFO] |  |  +- io.micronaut.servlet:micronaut-servlet-core:jar:4.7.0:compile
[INFO] |  |  |  \- io.micronaut:micronaut-http-server:jar:4.4.9:compile
[INFO] +- io.micronaut:micronaut-runtime:jar:4.5.1:compile
[INFO] |  +- io.micronaut:micronaut-discovery-core:jar:4.4.9:compile
[INFO] |  +- io.micronaut:micronaut-context-propagation:jar:4.4.9:compile
[INFO] |  +- io.micronaut:micronaut-core-reactive:jar:4.4.9:compile
[INFO] |  +- io.micronaut:micronaut-http:jar:4.4.9:compile
[INFO] |  \- io.micronaut:micronaut-retry:jar:4.4.9:compile
[INFO] +- io.micronaut:micronaut-http-client:jar:4.4.9:compile
[INFO] |  +- io.micronaut:micronaut-http-client-core:jar:4.4.9:compile
[INFO] |  +- io.micronaut:micronaut-websocket:jar:4.4.9:compile
[INFO] |  +- io.micronaut:micronaut-http-netty:jar:4.4.9:compile
[INFO] |  |  +- io.micronaut:micronaut-buffer-netty:jar:4.4.9:compile
[INFO] +- io.micronaut.testresources:micronaut-test-resources-client:jar:2.5.1:provided
[INFO] |  +- io.micronaut:micronaut-json-core:jar:4.4.9:compile
[INFO] |  \- io.micronaut.testresources:micronaut-test-resources-core:jar:2.5.2:provided
[INFO] +- io.micronaut.test:micronaut-test-junit5:jar:4.3.0:test
[INFO] |  \- io.micronaut.test:micronaut-test-core:jar:4.3.0:test

```

### Example Application

_No response_

### Version

4.4.3

--- a/test-resources-build-tools/src/main/java/io/micronaut/testresources/buildtools/ServerUtils.java
+++ b/test-resources-build-tools/src/main/java/io/micronaut/testresources/buildtools/ServerUtils.java
@@ -69,16 +69,20 @@ public class ServerUtils {
     private static final String SERVER_URI = "server.uri";
     private static final String SERVER_ACCESS_TOKEN_MICRONAUT_PROPERTY = "server.access-token";
     private static final String SERVER_ACCESS_TOKEN = "server.access.token";
+    private static final String ACCESS_TOKEN_HEADER = "Access-Token";
     private static final String SERVER_CLIENT_READ_TIMEOUT = "server.client.read.timeout";
     private static final String SERVER_IDLE_TIMEOUT_MINUTES = "server.idle.timeout.minutes";
     private static final String SERVER_ENTRY_POINT =
         "io.micronaut.testresources.server.TestResourcesService";
+    private static final String REQUIREMENTS_ENTRIES_PATH = "/requirements/entries";
     private static final String MICRONAUT_SERVER_PORT = "micronaut.server.port";
     private static final String JMX_SYSTEM_PROPERTY = "com.sun.management.jmxremote";
     private static final String CDS_HASH = "cds.bin";
     private static final String CDS_FILE = "cds.jsa";
     private static final String CDS_CLASS_LST = "cds.classlist";
     private static final String FLAT_JAR = "flat.jar";
+    private static final String JSON_CONTENT_TYPE = "application/json";
+    private static final int SERVER_PROBE_TIMEOUT_MS = 1000;
 
     // See io.micronaut.testresources.testcontainers.DockerSupport.TIMEOUT
     private static final String DOCKER_CHECK_TIMEOUT_SECONDS_ENV =
@@ -184,17 +188,39 @@ public class ServerUtils {
         Optional<ServerSettings> maybeServerSettings = readServerSettings(serverSettingsDirectory);
         if (maybeServerSettings.isPresent()) {
             LOGGER.info("Server settings found in {}", serverSettingsDirectory);
-            ServerSettings serverSettings = maybeServerSettings.get();
-            if (explicitPort != null && isServerStarted(explicitPort)) {
-                if (serverSettings.getPort() == explicitPort) {
+        }
+        if (explicitPort != null) {
+            ServiceProbeResult savedServerProbe = null;
+            if (maybeServerSettings.isPresent() && maybeServerSettings.get().getPort() == explicitPort) {
+                ServerSettings serverSettings = maybeServerSettings.get();
+                savedServerProbe = probeServer(serverSettings);
+                if (savedServerProbe.isReusable()) {
                     return serverSettings;
                 }
-                throw new IllegalStateException("Server already started on port " + explicitPort +
-                                                " but settings file says it should be on port " +
-                                                serverSettings.getPort());
             }
-            if (isServerStarted(serverSettings.getPort())) {
+            ServiceProbeResult explicitPortProbe =
+                probeServer(new ServerSettings(explicitPort, accessToken, clientTimeoutMs, serverIdleTimeoutMinutes));
+            if (explicitPortProbe.isReusable()) {
+                ServerSettings settings =
+                    new ServerSettings(explicitPort, accessToken, clientTimeoutMs, serverIdleTimeoutMinutes);
+                writeServerSettings(serverSettingsDirectory, settings);
+                return settings;
+            }
+            if (explicitPortProbe.isRunning()) {
+                throw explicitPortReuseFailure(explicitPort, explicitPortProbe.getFailureReason());
+            }
+            if (savedServerProbe != null && savedServerProbe.isRunning()) {
+                throw explicitPortReuseFailure(explicitPort, savedServerProbe.getFailureReason());
+            }
+        } else if (maybeServerSettings.isPresent()) {
+            ServerSettings serverSettings = maybeServerSettings.get();
+            ServiceProbeResult serverProbe = probeServer(serverSettings);
+            if (serverProbe.isReusable()) {
                 return serverSettings;
+            }
+            if (serverProbe.isRunning()) {
+                LOGGER.warn("Ignoring stale test resources server settings in {} because {}",
+                    serverSettingsDirectory, serverProbe.getFailureReason());
             }
         }
         if (Files.exists(portFilePath)) {
@@ -230,6 +256,118 @@ public class ServerUtils {
             new ServerSettings(port, accessToken, clientTimeoutMs, serverIdleTimeoutMinutes);
         writeServerSettings(serverSettingsDirectory, settings);
         return settings;
+    }
+
+    private static IllegalStateException explicitPortReuseFailure(int explicitPort, String failureReason) {
+        return new IllegalStateException("Explicit test resources port " + explicitPort + " is already in use "
+            + "by a service that could not be validated as Micronaut Test Resources: " + failureReason);
+    }
+
+    private static ServiceProbeResult probeServer(ServerSettings serverSettings) {
+        int port = serverSettings.getPort();
+        if (!isServerStarted(port)) {
+            return ServiceProbeResult.notRunning();
+        }
+        HttpURLConnection conn = null;
+        try {
+            URL url = new URL("http://localhost:" + port + REQUIREMENTS_ENTRIES_PATH);
+            conn = (HttpURLConnection) url.openConnection();
+            conn.setConnectTimeout(SERVER_PROBE_TIMEOUT_MS);
+            conn.setReadTimeout(SERVER_PROBE_TIMEOUT_MS);
+            conn.setRequestMethod("GET");
+            conn.setRequestProperty("Accept", JSON_CONTENT_TYPE);
+            String accessToken = serverSettings.getAccessToken().orElse(null);
+            if (accessToken != null) {
+                conn.setRequestProperty(ACCESS_TOKEN_HEADER, accessToken);
+            }
+            int responseCode = conn.getResponseCode();
+            if (responseCode == HttpURLConnection.HTTP_UNAUTHORIZED) {
+                return ServiceProbeResult.runningButInvalid("the saved access token was rejected");
+            }
+            if (responseCode != HttpURLConnection.HTTP_OK) {
+                return ServiceProbeResult.runningButInvalid("it responded with HTTP " + responseCode);
+            }
+            if (!isJsonContentType(conn.getContentType())) {
+                return ServiceProbeResult.runningButInvalid("it did not return JSON");
+            }
+            try (InputStream inputStream = conn.getInputStream()) {
+                String body = new String(inputStream.readAllBytes(), StandardCharsets.UTF_8);
+                if (!isJsonStringArray(body)) {
+                    return ServiceProbeResult.runningButInvalid("it did not return the expected JSON array payload");
+                }
+            }
+            return ServiceProbeResult.reusable();
+        } catch (IOException e) {
+            return ServiceProbeResult.runningButInvalid("probing failed with " + e.getClass().getSimpleName());
+        } finally {
+            if (conn != null) {
+                conn.disconnect();
+            }
+        }
+    }
+
+    private static boolean isJsonContentType(String contentType) {
+        return contentType != null && contentType.toLowerCase().contains(JSON_CONTENT_TYPE);
+    }
+
+    private static boolean isJsonStringArray(String body) {
+        int cursor = skipWhitespace(body, 0);
+        if (cursor >= body.length() || body.charAt(cursor) != '[') {
+            return false;
+        }
+        cursor = skipWhitespace(body, cursor + 1);
+        if (cursor < body.length() && body.charAt(cursor) == ']') {
+            return skipWhitespace(body, cursor + 1) == body.length();
+        }
+        while (cursor < body.length()) {
+            cursor = consumeJsonString(body, cursor);
+            if (cursor < 0) {
+                return false;
+            }
+            cursor = skipWhitespace(body, cursor);
+            if (cursor >= body.length()) {
+                return false;
+            }
+            char next = body.charAt(cursor);
+            if (next == ']') {
+                return skipWhitespace(body, cursor + 1) == body.length();
+            }
+            if (next != ',') {
+                return false;
+            }
+            cursor = skipWhitespace(body, cursor + 1);
+        }
+        return false;
+    }
+
+    private static int skipWhitespace(String body, int cursor) {
+        while (cursor < body.length() && Character.isWhitespace(body.charAt(cursor))) {
+            cursor++;
+        }
+        return cursor;
+    }
+
+    private static int consumeJsonString(String body, int cursor) {
+        if (cursor >= body.length() || body.charAt(cursor) != '"') {
+            return -1;
+        }
+        cursor++;
+        while (cursor < body.length()) {
+            char current = body.charAt(cursor);
+            if (current == '"') {
+                return cursor + 1;
+            }
+            if (current == '\\') {
+                cursor++;
+                if (cursor >= body.length()) {
+                    return -1;
+                }
+            } else if (Character.isISOControl(current)) {
+                return -1;
+            }
+            cursor++;
+        }
+        return -1;
     }
 
     /**
@@ -431,6 +569,42 @@ public class ServerUtils {
 
         default boolean isCDSDumpInvocation() {
             return false;
+        }
+    }
+
+    private static final class ServiceProbeResult {
+        private final boolean reusable;
+        private final boolean running;
+        private final String failureReason;
+
+        private ServiceProbeResult(boolean reusable, boolean running, String failureReason) {
+            this.reusable = reusable;
+            this.running = running;
+            this.failureReason = failureReason;
+        }
+
+        private static ServiceProbeResult reusable() {
+            return new ServiceProbeResult(true, true, null);
+        }
+
+        private static ServiceProbeResult notRunning() {
+            return new ServiceProbeResult(false, false, "it is not running");
+        }
+
+        private static ServiceProbeResult runningButInvalid(String failureReason) {
+            return new ServiceProbeResult(false, true, failureReason);
+        }
+
+        private boolean isReusable() {
+            return reusable;
+        }
+
+        private boolean isRunning() {
+            return running;
+        }
+
+        private String getFailureReason() {
+            return failureReason;
         }
     }
 

--- a/test-resources-build-tools/src/main/java/io/micronaut/testresources/buildtools/ServerUtils.java
+++ b/test-resources-build-tools/src/main/java/io/micronaut/testresources/buildtools/ServerUtils.java
@@ -299,7 +299,7 @@ public class ServerUtils {
                     return ServiceProbeResult.runningButInvalid("it did not return the expected JSON array payload");
                 }
             }
-            return ServiceProbeResult.reusable();
+            return ServiceProbeResult.reusableServer();
         } catch (IOException e) {
             return ServiceProbeResult.runningButInvalid("probing failed with " + e.getClass().getSimpleName());
         } finally {
@@ -575,18 +575,8 @@ public class ServerUtils {
         }
     }
 
-    private static final class ServiceProbeResult {
-        private final boolean reusable;
-        private final boolean running;
-        private final String failureReason;
-
-        private ServiceProbeResult(boolean reusable, boolean running, String failureReason) {
-            this.reusable = reusable;
-            this.running = running;
-            this.failureReason = failureReason;
-        }
-
-        private static ServiceProbeResult reusable() {
+    private record ServiceProbeResult(boolean reusable, boolean running, String failureReason) {
+        private static ServiceProbeResult reusableServer() {
             return new ServiceProbeResult(true, true, null);
         }
 

--- a/test-resources-build-tools/src/main/java/io/micronaut/testresources/buildtools/ServerUtils.java
+++ b/test-resources-build-tools/src/main/java/io/micronaut/testresources/buildtools/ServerUtils.java
@@ -42,6 +42,7 @@ import java.util.Collections;
 import java.util.HashMap;
 import java.util.HashSet;
 import java.util.List;
+import java.util.Locale;
 import java.util.Map;
 import java.util.Optional;
 import java.util.Properties;
@@ -282,7 +283,9 @@ public class ServerUtils {
             }
             int responseCode = conn.getResponseCode();
             if (responseCode == HttpURLConnection.HTTP_UNAUTHORIZED) {
-                return ServiceProbeResult.runningButInvalid("the saved access token was rejected");
+                return ServiceProbeResult.runningButInvalid(accessToken != null
+                    ? "the access token was rejected"
+                    : "an access token is required");
             }
             if (responseCode != HttpURLConnection.HTTP_OK) {
                 return ServiceProbeResult.runningButInvalid("it responded with HTTP " + responseCode);
@@ -307,7 +310,7 @@ public class ServerUtils {
     }
 
     private static boolean isJsonContentType(String contentType) {
-        return contentType != null && contentType.toLowerCase().contains(JSON_CONTENT_TYPE);
+        return contentType != null && contentType.toLowerCase(Locale.ROOT).contains(JSON_CONTENT_TYPE);
     }
 
     private static boolean isJsonStringArray(String body) {
@@ -422,7 +425,7 @@ public class ServerUtils {
             HttpURLConnection conn = (HttpURLConnection) url.openConnection();
             conn.setRequestMethod("POST");
             serverSettings.getAccessToken()
-                .ifPresent(token -> conn.setRequestProperty("Access-Token", token));
+                .ifPresent(token -> conn.setRequestProperty(ACCESS_TOKEN_HEADER, token));
             try (InputStream is = conn.getInputStream()) {
                 is.read();
             }

--- a/test-resources-build-tools/src/test/groovy/io/micronaut/testresources/buildtools/ServerUtilsTest.groovy
+++ b/test-resources-build-tools/src/test/groovy/io/micronaut/testresources/buildtools/ServerUtilsTest.groovy
@@ -1,7 +1,9 @@
 package io.micronaut.testresources.buildtools
 
+import com.sun.net.httpserver.HttpServer
 import io.micronaut.context.ApplicationContext
 import io.micronaut.http.annotation.Controller
+import io.micronaut.http.annotation.Get
 import io.micronaut.http.annotation.Post
 import io.micronaut.runtime.server.EmbeddedServer
 import jakarta.inject.Inject
@@ -9,6 +11,8 @@ import spock.lang.Specification
 import spock.lang.TempDir
 import spock.util.environment.RestoreSystemProperties
 
+import java.net.InetSocketAddress
+import java.nio.charset.StandardCharsets
 import java.nio.file.Files
 import java.nio.file.Path
 import java.util.concurrent.atomic.AtomicInteger
@@ -167,6 +171,69 @@ class ServerUtilsTest extends Specification {
 
         cleanup:
         applicationContext.stop()
+    }
+
+    def "starts a fresh server when persisted settings point to another service"() {
+        def portFile = tmpDir.resolve("port-file")
+        def settingsDir = tmpDir.resolve("settings")
+        def factory = Mock(ServerFactory)
+        def staleServer = startRequirementsServer(null, 'application/json', '{"service":"other"}')
+        def applicationContext = ApplicationContext.builder().start()
+        def embeddedServer = applicationContext.getBean(EmbeddedServer)
+        embeddedServer.start()
+        ServerUtils.writeServerSettings(settingsDir, new ServerSettings(staleServer.address.port, null, null, null))
+
+        when:
+        def settings = ServerUtils.startOrConnectToExistingServer(null, portFile, settingsDir, null, [], null, null, factory)
+
+        then:
+        1 * factory.startServer(_)
+        1 * factory.waitFor(_) >> {
+            portFile.toFile().text = "${embeddedServer.port}"
+        }
+        settings.port == embeddedServer.port
+
+        cleanup:
+        staleServer?.stop(0)
+        applicationContext.stop()
+    }
+
+    def "reuses existing server using the saved access token"() {
+        def portFile = tmpDir.resolve("port-file")
+        def settingsDir = tmpDir.resolve("settings")
+        def factory = Mock(ServerFactory)
+        def server = startRequirementsServer('secret-token')
+        ServerUtils.writeServerSettings(settingsDir, new ServerSettings(server.address.port, 'secret-token', null, null))
+
+        when:
+        def settings = ServerUtils.startOrConnectToExistingServer(null, portFile, settingsDir, null, [], null, null, factory)
+
+        then:
+        0 * factory.startServer(_)
+        0 * factory.waitFor(_)
+        settings == new ServerSettings(server.address.port, 'secret-token', null, null)
+
+        cleanup:
+        server?.stop(0)
+    }
+
+    def "fails clearly when an explicit port belongs to another service"() {
+        def portFile = tmpDir.resolve("port-file")
+        def settingsDir = tmpDir.resolve("settings")
+        def factory = Mock(ServerFactory)
+        def server = startRequirementsServer(null, 'text/plain', 'not test resources')
+
+        when:
+        ServerUtils.startOrConnectToExistingServer(server.address.port, portFile, settingsDir, null, [], null, null, factory)
+
+        then:
+        def ex = thrown(IllegalStateException)
+        ex.message == "Explicit test resources port ${server.address.port} is already in use by a service that could not be validated as Micronaut Test Resources: it did not return JSON"
+        0 * factory.startServer(_)
+        0 * factory.waitFor(_)
+
+        cleanup:
+        server?.stop(0)
     }
 
     def "supports class data sharing"() {
@@ -332,9 +399,34 @@ class ServerUtilsTest extends Specification {
         @Inject
         ApplicationContext ctx
 
+        @Get("/requirements/entries")
+        List<String> entries() {
+            ['stub.entry']
+        }
+
         @Post("/stop")
         void close() {
             ctx.close()
         }
+    }
+
+    private static HttpServer startRequirementsServer(String expectedAccessToken = null,
+                                                      String contentType = 'application/json',
+                                                      String body = '["stub.entry"]') {
+        def server = HttpServer.create(new InetSocketAddress("localhost", 0), 0)
+        server.createContext('/requirements/entries') { exchange ->
+            if (expectedAccessToken != null && exchange.requestHeaders.getFirst('Access-Token') != expectedAccessToken) {
+                exchange.sendResponseHeaders(401, -1)
+                exchange.close()
+                return
+            }
+            byte[] response = body.getBytes(StandardCharsets.UTF_8)
+            exchange.responseHeaders.set('Content-Type', contentType)
+            exchange.sendResponseHeaders(200, response.length)
+            exchange.responseBody.write(response)
+            exchange.close()
+        }
+        server.start()
+        server
     }
 }

--- a/test-resources-build-tools/src/test/groovy/io/micronaut/testresources/buildtools/ServerUtilsTest.groovy
+++ b/test-resources-build-tools/src/test/groovy/io/micronaut/testresources/buildtools/ServerUtilsTest.groovy
@@ -236,6 +236,25 @@ class ServerUtilsTest extends Specification {
         server?.stop(0)
     }
 
+    def "fails clearly when an explicit port requires an access token"() {
+        def portFile = tmpDir.resolve("port-file")
+        def settingsDir = tmpDir.resolve("settings")
+        def factory = Mock(ServerFactory)
+        def server = startRequirementsServer('secret-token')
+
+        when:
+        ServerUtils.startOrConnectToExistingServer(server.address.port, portFile, settingsDir, null, [], null, null, factory)
+
+        then:
+        def ex = thrown(IllegalStateException)
+        ex.message == "Explicit test resources port ${server.address.port} is already in use by a service that could not be validated as Micronaut Test Resources: an access token is required"
+        0 * factory.startServer(_)
+        0 * factory.waitFor(_)
+
+        cleanup:
+        server?.stop(0)
+    }
+
     def "supports class data sharing"() {
         def portFile = tmpDir.resolve("port-file")
         def settingsDir = tmpDir.resolve("settings")

